libraries/lib-basic-ui/BasicUI.h

```diff
@@ -0,0 +1,289 @@
+/*!********************************************************************
+
+Audacity: A Digital Audio Editor
+
+@file BasicUI.h
+@brief Toolkit-neutral facade for basic user interface services
+
+Paul Licameli
+
+**********************************************************************/
+#ifndef __AUDACITY_BASIC_UI__
+#define __AUDACITY_BASIC_UI__
+
+#include <functional>
+#include <memory>
+#include "Identifier.h"
+#include "Internat.h"
+
+namespace BasicUI {
+
+//! @name Types used in the Services interface
+//! @{
+
+using Action = std::function<void()>;
+
+//! Subclasses may hold information such as a parent window pointer for a dialog.
+/*! The default-constructed empty value of this base class must be accepted by overrides of methods of
+ Services */
+class BASIC_UI_API WindowPlacement {
+public:
+   WindowPlacement() = default;
+
+   //! Don't slice
+   WindowPlacement( const WindowPlacement& ) PROHIBITED;
+   //! Don't slice
+   WindowPlacement &operator=( const WindowPlacement& ) PROHIBITED;
+   virtual ~WindowPlacement();
+};
+
+enum class ErrorDialogType {
+   ModelessError,
+   ModalError,
+   ModalErrorReport, /*!< If error reporting is enabled, may give option to
+                      send; if not, then like ModalError
+                      */
+};
+
+//! Options for variations of error dialogs; the default is for modal dialogs
+struct ErrorDialogOptions {
+
+   ErrorDialogOptions() = default;
+   //! Non-explicit
+   ErrorDialogOptions( ErrorDialogType type ) : type{ type } {}
+
+   //! @name Chain-call style initializers
+   //! @{
+
+   ErrorDialogOptions &&ModalHelp( bool modalHelp_ ) &&
+   { modalHelp = modalHelp_; return std::move(*this); }
+
+   ErrorDialogOptions &&Log( std::wstring log_ ) &&
+   { log = std::move(log_); return std::move(*this); }
+
+   //! @}
+
+   //! Type of help dialog
+   ErrorDialogType type{ ErrorDialogType::ModalError };
```

Comment:
Why not use m prefix here?

Reply:
Well... ErrorDialogOptions isn't really a CLASS (needing private data members, having some important invariant property, and requiring mutative member functions to preserve that invariant).

Rather, it's just a POD (plain-old-data) type, grouping some bunch of dumb data together, and not requiring any special invariant relations to hold among the values of the members.

In cases like that, I relax the rule about the "m" prefix.